Add path planners

reiniscimurs opened the pull request:
Hi,

In my repo I implemented a couple of basic path planners taken from https://github.com/AtsushiSakai/PythonRobotics and adapted them for use directly in IR-SIM. This could also be contributed directly in the simulator repository here. Here is a PR for adding RRT, A* and PRM planners. There is also a test that should allow to test them see their execution. Works well on my side but do test the functionality. Let me know if any changes are required.

In the future I would probably add some more classical path planners though not all of them.

hanruihua commented:
Thank you for your great contributions. This is something I've always wanted to do, and I've also implemented some navigation algorithms in an older version of ir-sim that will be transitioned over in the future.

I plan to construct a navigation stack similar to ROS so that people can combine different modules to perform navigation tasks, including localization, mapping, planning, and control. Thus, I want each module to have a unified interface and type. For example, the interfaces of these classic path planners should read the map and generate the path. The mapping algorithm should process sensor data to generate the map, and the instance of Env can work as the ground truth.

I have added a map class in the world/map folder and implemented a get_map function in env_base. Could you please unify the interfaces of these planners so that they can read the map instance and generate trajectories? If there are any improvements or requirements for the map class, please let me know.

hanruihua commented on irsim/lib/path_planners/a_star.py:
```diff
@@ -0,0 +1,244 @@
+
+class AStarPlanner:
+
+    def __init__(self, env, resolution):
```
I suggest that the planner class read the map object returned by the simulator, which contains properties such as width, height, and resolution.

hanruihua commented on tests/test_path_planners.py:
```diff
@@ -0,0 +1,32 @@
+)
+def test_rrt_planner(planner, resolution):
+    env = irsim.make("test_collision_world.yaml", save_ani=False, full=False, display=False)
+    planner = planner(env, resolution)
```
the map can be obtained by `env.get_map`

reiniscimurs replied:
Updated to use the map from `env.get_map`

hanruihua commented on tests/test_path_planners.py:
```diff
@@ -0,0 +1,32 @@
+    planner = planner(env, resolution)
+    robot_info = env.get_robot_info()
+    robot_state = env.get_robot_state()
+    rx, ry = planner.planning(
```
Could you please return the trajectory type which is a list of waypoints (2*1 vector). And you can use the  `draw_trajectory` function to draw these trajectories.

hanruihua replied:
Could you please return the trajectory type which is a list of waypoints (2*1 vector). And you can use the  `draw_trajectory` function to draw these trajectories.

hanruihua commented on irsim/lib/path_planners/a_star.py:
```diff
@@ -0,0 +1,244 @@
+                + str(self.parent_index)
+            )
+
+    def planning(self, sx, sy, gx, gy, show_animation=True):
```
I suggest that the start and goal poses be of the NumPy array type, as this is more conducive to efficient matrix operations on large maps and more conducive to extension to higher dimensions.

reiniscimurs replied:
All planners expect np.arrays as input and x, y positions are extracted from that. This should allow to have unified inputs in the future where multiple dimensions could be used.

reiniscimurs commented:
> Thank you for your great contributions. This is something I've always wanted to do, and I've also implemented some navigation algorithms in an older version of ir-sim that will be transitioned over in the future.
> 
> I plan to construct a navigation stack similar to ROS so that people can combine different modules to perform navigation tasks, including localization, mapping, planning, and control. Thus, I want each module to have a unified interface and type. For example, the interfaces of these classic path planners should read the map and generate the path. The mapping algorithm should process sensor data to generate the map, and the instance of Env can work as the ground truth.
> 
> I have added a map class in the world/map folder and implemented a get_map function in env_base. Could you please unify the interfaces of these planners so that they can read the map instance and generate trajectories? If there are any improvements or requirements for the map class, please let me know. Thanks again for your support.

Thanks for the guidance here. I also think it makes a lot of sense that all of these planners are as unified as possible and hopefully in the future the stack can be expanded. It's a good idea to already have a map class here as a unifying interface. I updated the code to use that and return trajectories directly and updated tests for that as well. Let me know if other updates are required.

Btw is there a formatter for this repo (ruff, black, etc.)? I was looking for a reference to something like that but did not see anything like that in the toml file.

Also PRM uses scipy which would need to be added to requirements.

hanruihua commented:
I change the arguments `map` to be `env_map` to avoid conflicting with the built-in map function. And the `scipy` is added in the toml file. I use black to format this project, which is added in the requirements files now. 

I also have an idea to define a base planner class, however I am not sure it is suitable for all planners. It will be under consideration for the future.

Thank you for your efforts.
